Zmdernar/wildcard questions

Zachary-Dernar opened the pull request.

Zachary-Dernar committed:
Adds wildcard canned questions to kbs

currently wip: has a few bugs with the button interactions

Zachary-Dernar committed:
Handles question click separately

May still need to look into the networking of this implementation

Zachary-Dernar committed:
Merge branch 'master' into zmdernar/wildcard-questions

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/SearchContainer.js:
```diff
@@ -19,6 +20,27 @@ class SearchContainer extends Component {
   }
 
+  getFiveRandomQueries() {
+    const totalQueries = randomQueries.length;
```
why not 
1. make a copy of the array with `randomQueries.slice(0)`
1. use a for-loop to iterate over x times, x being the number of queries in the array
1. create a new array by removing the `Math.floor(Math.random() * randomQueries.length)` using `array.splice(index, 1)`
1. use the new array as your random queries

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/SearchContainer.js:
```diff
@@ -52,53 +78,54 @@ class SearchContainer extends Component {
+                  this.state.random_queries.map((query) => {
+                    return (
+                      <span>
+                        <button className='clickable-tab'
```
you wan to add a `type='button'` to prevent this from submitting the form

Zachary-Dernar replied:
resolved this by using an anonymous function instead

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+}
+
+.clickable-tabs {
+  display: table;
```
any way to use `display: flex` here?

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+
+.clickable-tab {
+  color: #ffffff;
+  border-style: solid;
```
you can define all of these border-* in one line: `border: medium solid #f0f0f0`, although not used to using `medium` for a width

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+}
+
+.clickable-tab:hover {
+  transition: 0.4s;
```
what is being transitioned? usually you define it like `transition: opacity 0.4s`

Zachary-Dernar replied:
everything in that block is transitioning in that timeframe. is that not the right way to do that?

maniax89 replied:
then you would do `transition: all 0.4s;` -> take a look at the documentation https://www.w3schools.com/cssref/css3_pr_transition.asp

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+}
+
+.tab-panels--tab-content {
+  background: transparent;
```
usually `background` alone implies shorthand, which would look like `background: #ffffff url("img_tree.png") no-repeat right top;`

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -49,17 +127,24 @@ button.random-query--button {
   border-color: #003C32;
 }
 
+button:focus {
```
should this be applied to all buttons or just the query buttons?

Zachary-Dernar replied:
Applying this to all buttons provides consistency across the demo. Also, these are the only buttons to date in the demo: there are only links and text searches otherwise.

maniax89 replied:
there are 2 other buttons -> "Retrieve Answers" and the "Show full answer"/"Collapse full answer". global styles are not the goal for modular components. if we want global styles, they should reside in the `App.css` file

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+}
+
+.base--radio:checked {
+  border-left-color: purple;
```
probably want to use the colors defined in the watson color guide https://watson-developer-cloud.github.io/ui-components/style.html#colors

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+}
+
+.tab-panels {
+  background: transparent;
```
usually shorthand should look like this: `background: #ffffff url("img_tree.png") no-repeat right top;`

Zachary-Dernar replied:
I'm not sure what this should look like if I'm just setting it to `transparent`

maniax89 replied:
`background-color: transparent;` also this is default -> https://www.w3schools.com/cssref/pr_background-color.asp

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,88 @@
+  bottom: 3rem;
+}
+
+.clickable-tab {
```
we might want to wait on the visual design specifics before going crazy on the tweaks here

maniax89 commented on client/knowledge_base_search/src/App.css:
```diff
@@ -29,7 +29,10 @@ body * + * {
+.base--a:hover {
+  color: white;
+  border-bottom-color: white;
+  },
```
remove this comma

maniax89 commented on client/knowledge_base_search/src/App.css:
```diff
@@ -29,7 +29,10 @@ body * + * {
 
-.base--a:hover,
+.base--a:hover {
+  color: white;
```
https://watson-developer-cloud.github.io/ui-components/style.html#colors

Zachary-Dernar committed:
Merge branch 'master' into zmdernar/wildcard-questions

Zachary-Dernar committed:
Finishes up design for wildcard canned questions v1

Pending on if we change the results page searchcontainer state,
this will be done if we are not collapsing the header.

Zachary-Dernar committed:
Merge branch 'master' of https://github.com/watson-developer-cloud/discovery-starter-kit into zmdernar/wildcard-questions

Zachary-Dernar committed:
Merge branch 'zmdernar/wildcard-questions' of https://github.com/watson-developer-cloud/discovery-starter-kit into zmdernar/wildcard-questions

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/SearchContainer.js:
```diff
@@ -43,62 +51,54 @@ class SearchContainer extends Component {
+                  this.state.random_queries.map((query, index) => {
+                    return (
+                      <span key={'questionButton-'+index}>
+                        <button className={query === this.state.search_input
```
probably should provide `disabled={this.props.isFetching}` so users can't spam the system

Zachary-Dernar replied:
Did this, just didn't do it in that exact spot so this isn't disappearing

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/SearchContainer.js:
```diff
@@ -19,12 +20,19 @@ class SearchContainer extends Component {
+  getFiveRandomQueries() {
+    const allQueries = randomQueries.slice(0);
+    let myQueries = [];
+    for(let i=0; i<5; i++) {
```
spacing

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,113 @@
+  border-radius: 0.5rem;
+  background-color: inherit;
+  padding: 1rem;
+  font-size: 0.9em;
```
use `rem`

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,113 @@
+
+.clickable-tabs {
+  display: flex;
+  border-spacing: 1rem;
```
i don't believe this does anything unless `display: table;`

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,113 @@
+}
+
+.tab-panels--tab:focus {
+  color: #fff !important;
```
you should just be able to add another class to the element that you want to override these properties instead of using `!important` -> css specificity https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity

maniax89 commented on client/knowledge_base_search/src/containers/SearchContainer/styles.css:
```diff
@@ -1,11 +1,113 @@
+}
+
+.container {
+  max-width: !65rem;
```
what?

maniax89 reviewed (commented):
i think the css needs cleaned up a bit - its hard to validate what is needed and what is not at this point

Zachary-Dernar committed:
Minimal code refactor

Some overlapping css stuff. Oops.

Zachary-Dernar committed:
removes csstransitiongroup from searchcontainer

it wasn't used anymore so a warning was appearing. oops.

Zachary-Dernar committed:
Removed unnecessary <span>'s and updated tests

SearchContainer's tests now match the new behavior
